src/monotonic_path_preview/mod.rs

```diff
@@ -0,0 +1,911 @@
+        let (mut res, mut tail) = split_head_and_tail(path)?;
+
+        let mut seen_paths = collections::BTreeSet::new();
+
+        // While there's still components left to check...
+        while tail.len() > 0 {
+            debug!(
+                "Monotonic path: {:?}, to check: {:?}",
+                res, tail
+            );
+
+            // Grab the component at the front of the tail, so we can check it.
+            let current = tail.pop_front()
+                .expect("len() > 0 but vec is empty?");
+
+            if &current == "." {
+                // A "." component can be ignored.
+                //
+                //    blah/./blah -> blah/blah
+
+            } else if &current == ".." {
+                // If we get a ParentDir component, the component at the end of
+                // `res` might be a symlink. in which case we'll have to splice
+                // the target path in at the beginning of `tail` so we have the
+                // full, monotonic path.
+                //
+                // If `foo` is not a symlink:
+                //
+                //    blah/foo/../blah -> blah/blah
+                //
+                // If `foo` is a symlink to `relative/target`:
+                //
+                //    blah/symlink/../blah -> blah/relative/target/../blah
+                //
+                // If `foo` is a symlink to `/absolute/target`:
+                //
+                //    blah/symlink/../blah -> /absolute/target/../blah
+                //
+
+                // But first, let's check for symlink loops. If we've already
+                // dereferenced this symlink before, we've hit a loop and we
+                // will never find a suitable answer.
+                if seen_paths.contains(&res) {
+                    return Err(Error::SymlinkLoops(res));
+                }
+
+                // This is the first time we've dereferenced this symlink, note
+                // it down.
+                seen_paths.insert(res.clone());
+
+                match read_link_if_exists(&res)? {
+                    // The component before the ParentDir component was not
+                    // a symlink, so we don't need to do anything to `tail`.
+                    None => (),
+
+                    // The component before the ParentDir component *was* a
+                    // symlink, so we need to add it to tail so we'll get around
+                    // to checking it.
+                    Some(target) => {
+                        debug!("Symlink target: {:?}", target);
+
+                        // We'll re-check this ".." component once we've
+                        // checked all the components from target. Since we're
+                        // pushing to the front, we're pushing components in
+                        // reverse order, and so we have to push this first.
+                        tail.push_front("..".into());
```

Comment:
this logic is confusing to me. We've got three moving pieces here that we are mutating: `tail`, `res` _and_ `target`. I'm having trouble keeping track of what we are trying to do.

Can this logic be simplified? Maybe a few functions that break it up and don't mutate multiple variables?

Reply:
The basic approach is pretty simple: we take an input path that we don't trust, we check each component and move it to the "trusted" pile, and once we've checked everything we return the trusted pile.

For example, let's take a path like "/foo/bar":

1. We split the path into the root ("/") and the tail ("foo", "bar")
1. We assume the root always exists, so it automatically goes into the `res`ult we're going to return.

        res: "/"
        tail: "foo", "bar"

1. We look at the first element in `tail`: "foo" looks reasonable, so we move it across to `res`, and now we have

        res: "/", "foo"
        tail: "bar"

1. We look at the first element in `tail`: "bar" looks reasonable, so we move it across to `res`, and now we have

        res: "/", "foo", "bar"
        tail:

1. `tail` is now empty, so we return `res`

That's a nice, straight-forward loop, easy to understand. However, things get more complex once ".." components enter the picture. Let's say we're given the path "/foo/../bar" where "/foo" is a symlink to "xyz":

1. As before, we split the path, check the components of `tail` and move them across to `res` if they look reasonable, until we get to

        res: "/", "foo"
        tail: "..", "bar"

1. The first component in `tail` is "..", which means the last component of `res` is immediately suspect, so we pop it off and dereference it, getting "xyz", which we shove onto `tail` so it will get checked in due course

        res: "/"
        tail: "xyz", "..", "bar"

1. "xyz" looks reasonable

        res: "/", "xyz"
        tail: "..", "bar"

1. The first component of `tail` is ".." again, so we pop off the last component of `res`, but this time it doesn't dereference to anything interesting

        res: "/"
        tail: "bar"

1. "bar" looks reasonable

        res: "/", "bar"
        tail:

1. `tail` is now empty, so we return `res`

Does that make sense? Things get more complex if a symlink target contains multiple new components, and even more complex if the target is an absolute path instead of a relative one. Rust's type-safety and manual memory management adds complexity on top of that, but the above is the general idea.

I'll think about how I might structure the code to make things more obvious.